cupy/cutensor.pyx

```diff
@@ -2,46 +2,72 @@ import numpy
 import warnings
 
 import cupy
-from cupy_backends.cuda.api import runtime
+
+from libc.stdint cimport intptr_t, uint32_t, uint64_t
+from cupy.core._carray cimport shape_t
+from cupy.core.core cimport ndarray
+from cupy.core cimport internal
+from cupy_backends.cuda.libs.cutensor cimport Handle
+from cupy_backends.cuda.libs.cutensor cimport TensorDescriptor
+from cupy_backends.cuda.libs.cutensor cimport ContractionDescriptor
+from cupy_backends.cuda.libs.cutensor cimport ContractionFind
+from cupy_backends.cuda.libs.cutensor cimport ContractionPlan
+
+from cupy.core cimport core
+from cupy.core cimport _reduction
+from cupy.cuda cimport device
+from cupy_backends.cuda.api cimport runtime
 from cupy_backends.cuda.libs cimport cutensor
-from cupy.cuda import device
-from cupy.core import _reduction
 
-_handles = {}
-_tensor_descriptors = {}
-_contraction_descriptors = {}
-_contraction_finds = {}
-_contraction_plans = {}
 
+cdef dict _handles = {}
+cdef dict _tensor_descriptors = {}
+cdef dict _contraction_descriptors = {}
+cdef dict _contraction_finds = {}
+cdef dict _contraction_plans = {}
+
+
+cdef class Mode(object):
 
-class Mode(object):
+    cdef:
+        object _array
+        readonly int ndim
+        readonly intptr_t data
 
     def __init__(self, mode):
-        self.mode = numpy.array(mode, dtype=numpy.int32)
-        assert self.mode.ndim == 1
+        self._array = numpy.array(mode, dtype=numpy.int32)
+        assert self._array.ndim == 1
+        self.ndim = self._array.size
+        self.data = self._array.ctypes.data
+
+    def __repr__(self):
+        return 'mode([' + ', '.join(self._array) + '])'
 
-    @property
-    def ndim(self):
-        return self.mode.size
 
-    @property
-    def data(self):
-        return self.mode.ctypes.data
+cdef class _Scalar(object):
+
+    cdef:
+        object _array
+        readonly intptr_t ptr
+
+    def __init__(self, value, dtype):
+        self._array = numpy.asarray(value, dtype=dtype)
+        self.ptr = self._array.ctypes.data
 
     def __repr__(self):
-        return 'mode([' + ', '.join(self.mode) + '])'
+        return self._array.item()
 
 
-def get_handle():
-    dev = device.get_device_id()
+cpdef Handle get_handle():
+    cdef int dev = device.get_device_id()
```

Comment:
I think you can define `cdef Handle handle` here to make it cdef'd.

Reply:
This interface has been exposed to the users, so it should be defined by `cpdef` for backward compatibility.

Reply:
However, I don't think the users want this interface, so we can drop `get_handle`, `get_cuda_dtype` and `get_cutensor_dtype` in v8. How do you think? (cc: @kmaehashi)

Reply:
I think it's ok to break these interfaces.

Reply:
Sorry, this is what I meant:
```python
cpdef Handle get_handle():
    cdef Handle handle
    # and the rest follows
```

Reply:
Fixed! Thanks!